src/Lean/Server/ImportCompletion.lean

```diff
@@ -0,0 +1,141 @@
+    importId.isMissing && importCmd.getTailPos?.isSome && completionPos == importCmd.getTailPos?.get! + ' '
+
+/-- Checks whether `completionPos` points at a free space in the header. -/
+def isImportCmdCompletionRequest (headerStx : Syntax) (completionPos : String.Pos) : Bool :=
+  ! headerStx[1].getArgs.any fun importStx => importStx.getArgs.any fun arg =>
+    arg.getPos?.isSome && arg.getTailPos?.isSome
+      && arg.getPos?.get! <= completionPos && completionPos <= arg.getTailPos?.get!
+
+def computePartialImportCompletions
+    (headerStx : Syntax)
+    (completionPos : String.Pos)
+    (availableImports : ImportTrie)
+    : Array Name := Id.run do
+  let some importStxToComplete := headerStx[1].getArgs.find? fun importStx => Id.run do
+      -- `partialTrailingDotStx` ≙ `("." ident)?`
+      let partialTrailingDotStx := importStx[3]
+      if ! partialTrailingDotStx.hasArgs then
+        return false
+      let trailingDot := partialTrailingDotStx[0]
+      let some tailPos := trailingDot.getTailPos?
+        | return false
+      return tailPos == completionPos
+    | return #[]
+  let importPrefixToComplete := importStxToComplete[2].getId
+
+  let completions : Array Name :=
+    availableImports.matchingToArray importPrefixToComplete
+      |>.map fun matchingAvailableImport =>
+        matchingAvailableImport.replacePrefix importPrefixToComplete Name.anonymous
+
+  let nonEmptyCompletions := completions.filter fun completion => !completion.isAnonymous
+
+  return nonEmptyCompletions.insertionSort (Name.cmp · · == Ordering.lt)
+
+def isImportCompletionRequest (text : FileMap) (headerStx : Syntax) (params : CompletionParams) : Bool :=
+  let completionPos := text.lspPosToUtf8Pos params.position
+  let headerStartPos := headerStx.getPos?.getD 0
+  let headerEndPos := headerStx.getTailPos?.getD headerStartPos
+  completionPos <= headerEndPos + ' ' + ' '
+
+def collectAvailableImportsFromLake : IO (Option AvailableImports) := do
+  let lakePath ← Lean.determineLakePath
+  let spawnArgs : IO.Process.SpawnArgs := {
+    stdin  := IO.Process.Stdio.null
+    stdout := IO.Process.Stdio.piped
+    stderr := IO.Process.Stdio.piped
+    cmd    := lakePath.toString
+    args   := #["available-imports"]
+  }
+  let lakeProc ← IO.Process.spawn spawnArgs
+  let stdout := String.trim (← lakeProc.stdout.readToEnd)
+  let exitCode ← lakeProc.wait
+  match exitCode with
+  | 0 =>
+    let Except.ok (availableImports : AvailableImports) ← pure (Json.parse stdout >>= fromJson?)
+      | throw <| IO.userError  s!"invalid output from `lake available-imports`:\n{stdout}"
+    return availableImports
+  | _ =>
+    return none
+
+def collectAvailableImportsFromSrcSearchPath : IO AvailableImports :=
+  (·.2) <$> StateT.run (s := #[]) do
+    let srcSearchPath ← initSrcSearchPath
+    for p in srcSearchPath do
+      if ! (← p.isDir) then
+        continue
+      Lean.forEachModuleInDir p fun mod => do
+        set <| (← get).push mod
```

Comment:
```suggestion
       modify (·.push mod)
```

Suggestion for code style that just happen to stick out to me while looking at this once more.

Reply:
Good point. Note that this is not just a code style issue but also a performance issue because `get` duplicates the array reference, preventing in-place updates of the array, whereas `modify` does not.